### PR TITLE
Engine Setup

### PR DESCRIPTION
**Pull request by anishvabardhan**

### CODE REVIEW

**Comment by anishvabardhan** [comment]
Old PR was getting updated so decided to open a new PR as Rahul Bhaiya suggested.

File: .gitignore
Comment (neortls007idev) [inline .gitignore:116]: Just remove this line rather than just removing the not symbol

File: Engine/Code/Engine/Graphics/Mesh.cpp
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Graphics/Mesh.cpp:6]: I don't see a use of vertex.h here anywhere not sure why you need the include

File: Engine/Code/Engine/Window/Window.cpp
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Window/Window.cpp]: could potentially pass these as default arguments to the function and override them in case you want to

File: ProtoGame2D/Code/Game/Framework/App.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/App.cpp:44]: Remove seed from time and move to an RNG class for random number generation stuff

File: .gitignore
Comment (rahulguptagamedev) [inline .gitignore]: Just remove this line rather than just removing the not symbol

File: Engine/Code/Engine/Graphics/Renderer.cpp
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Graphics/Renderer.cpp:261]: You don't need to recreate Vec2s just pass uvpos.m_Maxs same for positions since you initialize Zs in 2D to 0. SAme thing for the entire code block pretty much since you are doing copy constructs anyway

File: Engine/Code/Engine/Graphics/Vertex.h
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Graphics/Vertex.h:55]: @kevin30061994 Color aas floats and not uints ??

File: ProtoGame2D/Code/Game/Framework/App.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/App.cpp]: unsafe delete. I prefer to check my pointers. SAFE_DESTROY_PTR(x)

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp:68]: @anishvabardhan All this X stuff is test code to see if input system is working as you thought it would I am guessing?

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp]: don't create this every frame there is no need to create it once at construction time and use it unless this keeps changing

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp]: This should not be in the beginning of the frame move it all to render

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp:26]: Render function call hence move to render function of the game

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp]: Just a recommendation add constexpr colors to your Engine so that you can essentially just call something like Color::PURPLE

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (rahulguptagamedev) [inline ProtoGame2D/Code/Game/Framework/Game.cpp]: 1. other than swap buffers move all to render function.
2. no point in clearing right in end frame and and then clearing again in begin frame
3. Do not bind shaders that you are not using

**Review by rahulguptagamedev — CHANGES_REQUESTED** [review]

File: Engine/Code/Engine/Graphics/Mesh.cpp
Comment (anishvabardhan) [reply Engine/Code/Engine/Graphics/Mesh.cpp:6]: VertexBufferLayout is delcared in the vertex file.

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/Game.cpp:68]: yup just for testing

File: ProtoGame2D/Code/Game/Framework/App.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/App.cpp:44]: done

File: Engine/Code/Engine/Maths/Random.cpp
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Maths/Random.cpp]: Why is seed time based for all we care pick a default seed like 1234 and pass in a value in case you want to have a different seed on game reset could just be an integer on the app that you increment and pass to the seed on reset

File: Engine/Code/Engine/Core/Time.cpp
Comment (rahulguptagamedev) [inline Engine/Code/Engine/Core/Time.cpp:48]: Again time should not have a concept of seed at all.

File: Engine/Code/Engine/Maths/Random.cpp
Comment (anishvabardhan) [reply Engine/Code/Engine/Maths/Random.cpp]: done

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/Game.cpp]: done

File: Engine/Code/Engine/Window/Window.cpp
Comment (anishvabardhan) [reply Engine/Code/Engine/Window/Window.cpp:113]: done

Comment (anishvabardhan) [reply Engine/Code/Engine/Window/Window.cpp]: done

File: ProtoGame2D/Code/Game/Framework/App.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/App.cpp]: what do u mean i am initialising as a nullptr after deleting it

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/Game.cpp]: done

File: .gitignore
Comment (anishvabardhan) [reply .gitignore]: done

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/Game.cpp]: done

File: Engine/Code/Engine/Graphics/Renderer.cpp
Comment (knappoly) [reply Engine/Code/Engine/Graphics/Renderer.cpp:261]: If it were up to me I would just have a Color3f that takes individual components and do the same for the others as well. @neortls007idev : This assumption only works if the constructor is not marked "explicit" which in this case it would be since I recall asking him to make constructors explicit.

Comment (knappoly) [reply Engine/Code/Engine/Graphics/Renderer.cpp:261]: But in this case its up to @anishvabardhan to decide. Not my engine.

File: Engine/Code/Engine/Graphics/Vertex.h
Comment (knappoly) [reply Engine/Code/Engine/Graphics/Vertex.h:55]: Well thats fine as long as the vertex buffer elments are created as such

Comment (knappoly) [reply Engine/Code/Engine/Graphics/Vertex.h:55]: I don't think he has a color struct yet to do the uint split

File: ProtoGame2D/Code/Game/Framework/App.cpp
Comment (knappoly) [reply ProtoGame2D/Code/Game/Framework/App.cpp]: You should always do a check for the pointer for stuff like this: 
if (m_Game)
{ 
   delete m_Game;
  m_game = null;
}

Comment (knappoly) [reply ProtoGame2D/Code/Game/Framework/App.cpp]: There may be instances that code could avoid creating the game instance coz of some early exit and crash because of this

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (knappoly) [reply ProtoGame2D/Code/Game/Framework/Game.cpp:26]: Did we not have a conversation bout Renderer owning shader instances? Why are we newing off a ShaderDefinition object here?

File: Engine/Code/Engine/Graphics/Renderer.cpp
Comment (knappoly) [inline Engine/Code/Engine/Graphics/Renderer.cpp]: I thought I told you before that this function is not named right? You can't draw a framebuffer .. What you're doing here is DrawMeshImmediate

Comment (knappoly) [reply Engine/Code/Engine/Graphics/Renderer.cpp]: Or if I'm ready currect its drawing a full screen quad .. So the name would be DrawFullScreenQuad

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (knappoly) [inline ProtoGame2D/Code/Game/Framework/Game.cpp]: The intent here shouldn't need window dimensions. A full screen quad is always  a quad that extends from 0,0 to 1,1 since its aalways going to be treated as NDC space. You shouldn't be trying to multiply the matrices, vertex shader should just pass the verts as is .. Or for that matter you don't even need a vertex buffer bound for this.

File: ProtoGame2D/Code/Game/Framework/Game.cpp
Comment (anishvabardhan) [reply ProtoGame2D/Code/Game/Framework/Game.cpp:26]: Done

File: Engine/Code/Engine/Graphics/Renderer.cpp
Comment (anishvabardhan) [reply Engine/Code/Engine/Graphics/Renderer.cpp]: Done

File: Engine/Code/Engine/Graphics/Renderer.cpp
Comment (anishvabardhan) [reply Engine/Code/Engine/Graphics/Renderer.cpp:261]: The constructors are explicit

File: Engine/Code/Engine/Graphics/Renderer.h
Comment (knappoly) [inline Engine/Code/Engine/Graphics/Renderer.h:92]: You shouldn't need these params in the first place. We can change this after this merge.

**Review by knappoly — APPROVED** [review]

**Review by rahulguptagamedev — APPROVED** [review]

**Pull request merged**
